pkg/helper/openshift.go

```diff
@@ -0,0 +1,126 @@
+package helper
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	configv1 "github.com/openshift/api/config/v1"
+	"golang.org/x/mod/semver"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func GetOpenshiftVersion(ctx context.Context, client client.Client) (string, bool, error) {
+	clusterVersion := &configv1.ClusterVersion{}
+
+	if err := client.Get(ctx, types.NamespacedName{
+		Name: "version",
+	}, clusterVersion); err != nil {
+		if errors.IsNotFound(err) {
+			return "", false, nil
+		}
+
+		return "", false, err
+	}
+
+	return clusterVersion.Status.Desired.Version, true, nil
+}
+
+func CompareOpenshiftVersion(ctx context.Context, client client.Client, version string) (int, bool, error) {
+	currentVersion, ok, err := GetOpenshiftVersion(ctx, client)
+	if !ok || err != nil {
+		return 0, ok, err
+	}
+
+	return semver.Compare(fmt.Sprintf("v%s", currentVersion), fmt.Sprintf("v%s", version)), true, nil
+}
+
+func SumRateForOpenshiftVersion(ctx context.Context, client client.Client) (string, error) {
+	sumRate := "sum_irate"
+
+	// Compare the current Openshft version to 4.9
+	comparison, ok, err := CompareOpenshiftVersion(ctx, client, "4.9")
+	if err != nil {
+		return "", err
+	}
+	// If the version could not be found, return the default mutation
+	if !ok {
+		return sumRate, nil
+	}
+
+	// If the version is less than 4.9, use sum_rate
+	if comparison < 0 {
+		sumRate = "sum_rate"
+	}
+
+	return sumRate, nil
+}
+
+func SumRateTemplateDataMutation(ctx context.Context, client client.Client) (TemplateDataMutation, error) {
+	sumRate, err := SumRateForOpenshiftVersion(ctx, client)
+	if err != nil {
+		return nil, err
+	}
+
+	return AddSumRateField(sumRate), nil
+}
+
+// AddSumRateField creates a templateDataMutation that constructs a struct
+// identical to the original with the added `SumRate` field, containing the
+// value of sumRate
+func AddSumRateField(sumRate string) TemplateDataMutation {
```

Comment:
I understand the motivation of this implementation based on the `reflect` package. 

I am ok with it, but I would like to share that IMHO the complexity added to solve a relatively simple use case does not pay off. 

Some (if not all) the dashboard templates have a new parameter called `SumRate`. I would add them to the data struct and the value passed to the `SumRate` will be computed based on the logic you implemented. 

For example, in `pkg/3scale/amp/component/generic_monitoring.go`
```go

func KubernetesResourcesByNamespaceGrafanaDashboard(sumRate string, ns string, appLabel string) *grafanav1alpha1.GrafanaDashboard {
	data := &struct {
		Namespace string
                SumRate string
	}{
		ns, sumRate,
	}
	return &grafanav1alpha1.GrafanaDashboard{
		ObjectMeta: metav1.ObjectMeta{
			Name: "kubernetes-resources-by-namespace",
			Labels: map[string]string{
				"monitoring-key": common.MonitoringKey,
				"app":            appLabel,
			},
		},
		Spec: grafanav1alpha1.GrafanaDashboardSpec{
			Json: assets.TemplateAsset("monitoring/kubernetes-resources-by-namespace-grafana-dashboard-1.json.tpl", data),
			Name: fmt.Sprintf("%s/kubernetes-resources-by-namespace-grafana-dashboard-1.json", ns),
		},
	}
}
```

The caller would be:
```go
sumRate, err := helper.SumRateForOpenshiftVersion(r.Context(), r.Client())  
if err != nil {                                                             
    return reconcile.Result{}, err                                          
}                  

grafanaDashboard := component.KubernetesResourcesByNamespaceGrafanaDashboard(sumRate, r.apiManager.Namespace, *r.apiManager.Spec.AppLabel)
err = r.ReconcileGrafanaDashboard(grafanaDashboard, reconcilers.GenericGrafanaDashboardsMutator)   
if err != nil {                                                                                                  
    return reconcile.Result{}, err                                                                               
}                                                                                                                
                                                

```

Reply:
Thanks @eguzki I followed your suggestion and simplify the inclusion of the `SumRate` parameter into the templates